mm/mm_heap/mm_free.c

```diff
@@ -78,34 +78,37 @@ static void add_delaylist(FAR struct mm_heap_s *heap, FAR void *mem)
  *
  ****************************************************************************/
 
-void mm_delayfree(FAR struct mm_heap_s *heap, FAR void *mem, bool delay)
+void mm_delayfree(FAR struct mm_heap_s *heap, FAR void *mem,
+                  bool delay, bool cleanup)
 {
   FAR struct mm_freenode_s *node;
   FAR struct mm_freenode_s *prev;
   FAR struct mm_freenode_s *next;
   size_t nodesize;
   size_t prevsize;
 
-  if (mm_lock(heap) < 0)
+  if (cleanup)
```

Comment:
mm_lock can't be replaced by cleanup because the memory block is filled too early if the current thread tries to free its stack.